samples/Matty/Matty.Server/Controllers/AuthorizationController.cs

```diff
@@ -0,0 +1,235 @@
+            var user = await _userManager.GetUserAsync(User) ??
+                throw new InvalidOperationException("The user details cannot be retrieved.");
+
+            // Retrieve the claims principal associated with the user code.
+            var result = await HttpContext.AuthenticateAsync(OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
+            if (result.Succeeded)
+            {
+                var principal = await _signInManager.CreateUserPrincipalAsync(user);
+
+                // Note: in this sample, the granted scopes match the requested scope
+                // but you may want to allow the user to uncheck specific scopes.
+                // For that, simply restrict the list of scopes before calling SetScopes.
+                principal.SetScopes(result.Principal.GetScopes());
+                principal.SetResources(await _scopeManager.ListResourcesAsync(principal.GetScopes()).ToListAsync());
+
+                foreach (var claim in principal.Claims)
+                {
+                    claim.SetDestinations(GetDestinations(claim, principal));
+                }
+
+                var properties = new AuthenticationProperties
+                {
+                    // This property points to the address OpenIddict will automatically
+                    // redirect the user to after validating the authorization demand.
+                    RedirectUri = "/"
+                };
+
+                return SignIn(principal, properties, OpenIddictServerAspNetCoreDefaults.AuthenticationScheme);
+            }
+
+            // Redisplay the form when the user code is not valid.
+            return View(new VerifyViewModel
+            {
+                Error = Errors.InvalidToken,
+                ErrorDescription = "The specified user code is not valid. Please make sure you typed it correctly."
+            });
+        }
+
+        [Authorize, FormValueRequired("submit.Deny")]
+        [HttpPost("~/connect/verify"), ValidateAntiForgeryToken]
+        // Notify OpenIddict that the authorization grant has been denied by the resource owner.
+        public IActionResult VerifyDeny() => Forbid(
+            authenticationSchemes: OpenIddictServerAspNetCoreDefaults.AuthenticationScheme,
+            properties: new AuthenticationProperties()
+            {
+                // This property points to the address OpenIddict will automatically
+                // redirect the user to after rejecting the authorization demand.
+                RedirectUri = "/"
+            });
+
+        [HttpPost("~/connect/token"), Produces("application/json")]
+        public async Task<IActionResult> Exchange()
+        {
+            var request = HttpContext.GetOpenIddictServerRequest() ??
+                throw new InvalidOperationException("The OpenID Connect request cannot be retrieved.");
+
+            if (request.IsAuthorizationCodeGrantType() || request.IsDeviceCodeGrantType() || request.IsRefreshTokenGrantType())
```

Comment:
Since the sample doesn't support the code flow, consider removing the `request.IsAuthorizationCodeGrantType()` check.